### PR TITLE
V22F-742 speed dreams Linux update

### PR DESCRIPTION
**Pull request by KungFuDonkey**
This PR transforms speed dreams to also work on linux. This PR will not include the github actions for linux and those will be done in another PR

### CODE REVIEW

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/frontend/IndicatorConfig.cpp [hunk 1]
```diff
@@ -3,23 +3,21 @@
...
+#include <sys/stat.h>
+#endif
 
 #include "IndicatorConfig.h"
-#include "Mediator.h"
 
 /// @brief        Loads the indicator data of every intervention action from config file in the given path
 /// @param p_path The path to the XML file containing the indicator data to load
-void IndicatorConfig::LoadIndicatorData(const char* p_path)
+void IndicatorConfig::LoadIndicatorData(const char* p_path, InterventionType p_interventionType)
```
Comment (shoekzema) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/frontend/IndicatorConfig.cpp:17]: add p_interventionType to documentation

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/rppUtils/RppUtils.hpp [hunk 2]
```diff
@@ -261,26 +271,81 @@ inline bool GetSdaFolder(std::experimental::filesystem::path& p_sdaFolder)
...
+{
+    std::string fullCommand = "gnome-terminal -- " + p_executablePath + " " + std::string(p_args);
+    popen(fullCommand.c_str(),"r");
+}
+
+/// @brief                       Starts a process from which you can also get the process handle
+/// @param  p_executablePath     The path to the executable
+/// @param  p_args               The arguments for the executable
+/// @param  p_processInformation The information about the process, this contains the handles
+inline void StartProcess(const std::string& p_executablePath, const char* p_args, PROCESS_INFORMATION& p_processInformation, const std::string& p_workingDirectory)
```
Comment (shoekzema) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/rppUtils/RppUtils.hpp:292]: missing p_workingDirectory in documentation

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/rppUtils/RppUtils.hpp [hunk 3]
```diff
@@ -261,26 +271,81 @@ inline bool GetSdaFolder(std::experimental::filesystem::path& p_sdaFolder)
...
+
+/// @brief                Execute a command in the CLI
+/// @param  p_command     The command
+/// @param  p_showCommand Whether to show the output of the command
+inline void ExecuteCLI(const char* p_command, bool p_showCommand)
+{
+    system(p_command);
+}
+
+inline bool GetSdaFolder(filesystem::path& p_sdaFolder)
```
Comment (shoekzema) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/rppUtils/RppUtils.hpp:306]: missing documentation

**Review by shoekzema — CHANGES_REQUESTED** [review]

File: speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/ResearcherMenu.cpp [hunk 4]
```diff
@@ -518,19 +518,22 @@ static void SelectBlackBox(void* /* dummy */)
...
     }
+
+#ifdef WIN32
     // Enforce that file ends in .exe
     if (std::strcmp(path.extension().string().c_str(), ".exe") != 0)
     {
         GfuiLabelSetText(s_scrHandle, m_errorLabel, MSG_ERROR_BLACK_BOX_NOT_EXE);
         return;
     }
+#endif
```
Comment (larsdekwant) [inline speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/ResearcherMenu.cpp:536]: Why is this sanity check only ncecessary for windows?

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backend/FileDataStorage.h [hunk 5]
```diff
@@ -46,7 +46,17 @@ class FileDataStorage
...
-    std::experimental::filesystem::path Initialize(
+    /// @brief Add the new value to the array in the correct compression step
+    /// @param p_values Array with values from the current compression step
+    /// @param p_value The new value of this time step for the variable
+    /// @param p_compressionStep The current compression step
+    template <typename TNumber>
+    void AddToArray(TNumber* p_values, TNumber p_value, int p_compressionStep) const
+    {
+        p_values[p_compressionStep] = p_value;
+    }
```
Comment (larsdekwant) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backend/FileDataStorage.h:57]: Why was this definition moved from the cpp to the header file?

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/mocks/DecisionMakerMock.h [hunk 6]
```diff
@@ -24,8 +24,8 @@ class DecisionMakerMock
         Situation = p_initialSituation;
         Track = p_track;
         BlackboxExecutablePath = p_blackBoxExecutablePath;
-        Recorder = p_recorder;
-        BlackBoxData = p_testSituations;
+        Record = p_recorder;
```
Comment (larsdekwant) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/mocks/DecisionMakerMock.h:27]: Why was this renamed??? I find Recorder to be a clearer name.

**Review by larsdekwant — CHANGES_REQUESTED** [review]

File: speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/ResearcherMenu.cpp (hunk 4 above)
Comment (KungFuDonkey) [reply speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/ResearcherMenu.cpp:536]: bc linux does not use .exe but has no extensions for executables

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backend/FileDataStorage.h (hunk 5 above)
Comment (KungFuDonkey) [reply speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backend/FileDataStorage.h:57]: It did not want to build on Linux as the definition of AddToArray is not available for tests, I was not able to use an explicit definition in the GCC compiler so I converted it to an implicit definition

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/mocks/DecisionMakerMock.h (hunk 6 above)
Comment (KungFuDonkey) [reply speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/mocks/DecisionMakerMock.h:27]: GCC does not allow class and variable names to be the same thing

**Review by larsdekwant — APPROVED** [review]

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backend/CMakeLists.txt [hunk 7]
```diff
@@ -41,8 +40,7 @@ SET(_HEADERS
 	IDecision.h
 	InterventionExecutor.h
 	InterventionExecutorAlwaysIntervene.h
-	InterventionExecutorAskFor.h
-	InterventionExecutorIndication.h
+        InterventionExecutorIndication.h
```
Comment (shoekzema) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backend/CMakeLists.txt:43]:
```suggestion
    InterventionExecutorIndication.h
```

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/unitTestUtilitiesTests/UtilsTests.cpp [hunk 8]
```diff
@@ -142,4 +142,49 @@ TEST(UtilsTests, PairWiseCoverageTest)
...
+void PairWise5(bool p_1, bool p_2, bool p_3, bool p_4, bool p_5) {SUCCEED() << p_1 << " " << p_2 << " " << p_3 << " " << p_4 << " " << p_5;}
+void PairWise6(bool p_1, bool p_2, bool p_3, bool p_4, bool p_5, bool p_6) {SUCCEED() << p_1 << " " << p_2 << " " << p_3 << " " << p_4 << " " << p_5 << " " << p_6;}
+void PairWise7(bool p_1, bool p_2, bool p_3, bool p_4, bool p_5, bool p_6, bool p_7) {SUCCEED() << p_1 << " " << p_2 << " " << p_3 << " " << p_4 << " " << p_5 << " " << p_6 << " " << p_7;}
+void PairWise8(bool p_1, bool p_2, bool p_3, bool p_4, bool p_5, bool p_6, bool p_7, bool p_8) {SUCCEED() << p_1 << " " << p_2 << " " << p_3 << " " << p_4 << " " << p_5 << " " << p_6 << " " << p_7 << " " << p_8;}
+
+bool bools[2] {true, false};
+
+TEST(UtilsTests, PairWiseTest2)
+{
+    PairWiseTest(PairWise2,bools,2,bools,2);
```
Comment (shoekzema) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/unitTestUtilitiesTests/UtilsTests.cpp:159]: Run Clang-format in the whole file

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/unitTestUtilitiesTests/UtilsTests.cpp [hunk 9]
```diff
@@ -142,4 +142,49 @@ TEST(UtilsTests, PairWiseCoverageTest)
...
     }
+}
+
+void PairWise2(bool p_1, bool p_2) {SUCCEED() << p_1 << " " << p_2;}
+void PairWise3(bool p_1, bool p_2, bool p_3) {SUCCEED() << p_1 << " " << p_2 << " " << p_3;}
+void PairWise4(bool p_1, bool p_2, bool p_3, bool p_4) {SUCCEED() << p_1 << " " << p_2 << " " << p_3 << " " << p_4;}
+void PairWise5(bool p_1, bool p_2, bool p_3, bool p_4, bool p_5) {SUCCEED() << p_1 << " " << p_2 << " " << p_3 << " " << p_4 << " " << p_5;}
+void PairWise6(bool p_1, bool p_2, bool p_3, bool p_4, bool p_5, bool p_6) {SUCCEED() << p_1 << " " << p_2 << " " << p_3 << " " << p_4 << " " << p_5 << " " << p_6;}
+void PairWise7(bool p_1, bool p_2, bool p_3, bool p_4, bool p_5, bool p_6, bool p_7) {SUCCEED() << p_1 << " " << p_2 << " " << p_3 << " " << p_4 << " " << p_5 << " " << p_6 << " " << p_7;}
+void PairWise8(bool p_1, bool p_2, bool p_3, bool p_4, bool p_5, bool p_6, bool p_7, bool p_8) {SUCCEED() << p_1 << " " << p_2 << " " << p_3 << " " << p_4 << " " << p_5 << " " << p_6 << " " << p_7 << " " << p_8;}
```
Comment (shoekzema) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/unitTestUtilitiesTests/UtilsTests.cpp:153]: No need to add documentation for all of them, but at least add it to the first one

**Review by shoekzema — CHANGES_REQUESTED** [review]

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/unitTestUtilitiesTests/UtilsTests.cpp [hunk 10]
```diff
@@ -142,4 +142,49 @@ TEST(UtilsTests, PairWiseCoverageTest)
...
+}
+
+TEST(UtilsTests, PairWiseTest7)
+{
+    PairWiseTest(PairWise7, bools,2,bools,2,bools,2,bools,2,bools,2,bools,2,bools,2);
+}
+
+TEST(UtilsTests, PairWiseTest8)
+{
+    PairWiseTest(PairWise8, bools,2, bools,2,bools,2,bools,2,bools,2,bools,2,bools,2,bools,2);
```
Comment (larsdekwant) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/unitTestUtilitiesTests/UtilsTests.cpp:191]: Add spaces after commas

**Review by larsdekwant — CHANGES_REQUESTED** [review]

**Review by shoekzema — APPROVED** [review]

**Review by suppergerrie2 — CHANGES_REQUESTED** [review]

**Review by suppergerrie2 — APPROVED** [review]
